core/services/ocrcommon/data_source.go

```diff
@@ -176,19 +194,94 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
 
 // Observe without saving to DB
 func (ds *inMemoryDataSource) Observe(ctx context.Context, timestamp ocr2types.ReportTimestamp) (*big.Int, error) {
-	_, finalResult, err := ds.executeRun(ctx, ObservationTimestamp{
+	_, trrs, err := ds.executeRun(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	finalResult := trrs.FinalResult(ds.lggr)
+	setEATelemetry(ds, trrs, finalResult, ObservationTimestamp{
 		Round:        timestamp.Round,
 		Epoch:        timestamp.Epoch,
 		ConfigDigest: timestamp.ConfigDigest.Hex(),
 	})
-	if err != nil {
-		return nil, err
-	}
+
 	return ds.parse(finalResult)
 }
 
+// inMemoryDataSourceCache is a time based cache wrapper for inMemoryDataSource.
+// If cache update is overdue Observe defaults to standard inMemoryDataSource behaviour.
+type inMemoryDataSourceCache struct {
+	cacheExpiration time.Duration
+	mu              *sync.RWMutex
+	lastUpdate      atomic.Pointer[time.Time]
+	latestTrrs      pipeline.TaskRunResults
+	latestResult    pipeline.FinalResult
+	*inMemoryDataSource
+}
+
+// isUpdateOverdue atomically checks if update is overdue.
+func (ds *inMemoryDataSourceCache) isUpdateOverdue() bool {
+	timeSinceUpdate := time.Since(*ds.lastUpdate.Load())
+	return timeSinceUpdate >= ds.cacheExpiration
+}
+
+// updater periodically updates data source cache.
+func (ds *inMemoryDataSourceCache) updater() {
+	ticker := time.NewTicker(ds.cacheExpiration)
+	for range ticker.C {
+		// arbitrary timeout
+		ctx, cancel := context.WithTimeout(context.Background(), time.Minute*1)
+		if err := ds.updateCache(ctx); err != nil {
+			ds.lggr.Warnw("failed to update cache", "err", err)
+		}
+		cancel()
+	}
+}
+
+func (ds *inMemoryDataSourceCache) updateCache(ctx context.Context) error {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+	if ds.isUpdateOverdue() {
+		_, trrs, err := ds.executeRun(ctx)
+		if err != nil {
+			return errors.Wrapf(err, "error executing run for spec ID %v", ds.spec.ID)
+		}
+		lastUpdate := time.Now()
+
+		ds.latestResult = trrs.FinalResult(ds.lggr)
+		ds.latestTrrs = trrs
+		ds.lastUpdate.Store(&lastUpdate)
+	}
+	return nil
+}
+
+func (ds *inMemoryDataSourceCache) Observe(ctx context.Context, timestamp ocr2types.ReportTimestamp) (*big.Int, error) {
+	// if cache update is overdue revert to standard inMemoryDataSource Observe behaviour
+	result := make(chan error)
+	go func() { result <- ds.updateCache(ctx) }()
```

Comment:
It looks like you might end up with lock contention on the `rs.mu` mutex. There is a `Lock` in `updateCache` and a subsequent `RLock` in `executeRun` for `currentAnswer`.

The docs say about mutexes: `It should not be used for recursive read locking; a blocked Lock call excludes new readers from acquiring the lock. See the documentation on the RWMutex type.`

Reply:
Its not an issue, it will just wait till the update is done and have the freshest data. In real world its almost imossible to happen as it either updated the cache right before or the cache was already up to date, which means that this can only happen if there was an update due in 1 milisecond and the lock got grabbed right before RLock() locked the read. 
tldr There are no race conditions, just a small chance of observe waiting for a fresh update which is how observe works currently for juels fee per coin anyway

Reply:
This is a readLock because I am looking if there is an easy way to make this cache work for the whole node instead of per job